core/blockchain.go

```diff
@@ -1736,3 +1743,47 @@ func (bc *BlockChain) SubscribeLogsEvent(ch chan<- []*types.Log) event.Subscript
 func (bc *BlockChain) SubscribeBlockProcessingEvent(ch chan<- bool) event.Subscription {
 	return bc.scope.Track(bc.blockProcFeed.Subscribe(ch))
 }
+
+// GetAncestorHash return the hash of ancestor at the given number
+func (bc *BlockChain) GetAncestorHash(ref *types.Header, target uint64) common.Hash {
+	number := ref.Number.Uint64() - 1
+	hash := ref.ParentHash
+	if target == number {
+		return hash
+	}
+	if target > number {
+		// Should never happen
+		log.Error("Ancestor number must be <= descendant", "ancestor target", target, "descendant", number)
```

Comment:
It seems like this code base's convention is to use `panic` for these scenarios. If it were me, I would have introduced a second `error` return value, but that's not the way this code base handles error conditions like these.

Reply:
Well, it's a caller error if so. The `blockhash` opcode has internal guards to prevent letting through a too-recent number, but I don't want to place a `panic` there -- if it's used from the p2p layer, I don't want a flaw in that code to crash the client.

Reply:
Right. It may be worth adding a comment to this method to make it explicit that this method is designed solely for `BLOCKHASH` so future contributors know the scope of this method.

Reply:
Well, it can be used for other purposes, however, you can't ask "give me the ancestor at `105` for ` [block 100, hash H]` and expect anything other than zeroes....